server/chat-commands.js

```diff
@@ -3863,59 +3863,56 @@ const commands = {
+		}
+		battle.allowExtraction[targetUser.userid].add(user.userid);
+		this.addModAction(`${user.name} consents to sharing battle team and choices with ${targetUser.name}.`);
+		if (Object.keys(battle.playerTable).length === battle.allowExtraction[targetUser.userid].size) {
+			this.addModAction(`${targetUser.name} has extracted the battle input log.`);
+			const inputLog = battle.inputLog.map(Chat.escapeHTML).join(`<br />`);
+			targetUser.sendTo(room, `|html|<div class="chat"><code style="white-space: pre-wrap; overflow-wrap: break-word; display: block">${inputLog}</code></div>`);
 		}
-		this.addModAction(`${user.userid} consents to sharing battle team and choices with ${targetUser.userid}.`);
-		if (battle.allowExtraction.join(',') !== `${targetUser.userid},${targetUser.userid}`) return;
-
-		this.addModAction(`${targetUser.name} has extracted the battle input log.`);
-		const inputLog = battle.inputLog.map(Chat.escapeHTML).join(`<br />`);
-		targetUser.sendTo(room, `|html|<div class="chat"><code style="white-space: pre-wrap; overflow-wrap: break-word; display: block">${inputLog}</code></div>`);
 	},
 
 	requestinputlog: 'exportinputlog',
 	exportinputlog(target, room, user) {
 		const battle = room.battle;
-		if (!battle) return this.errorReply(`This command only works in battle rooms.`);
+		if (!battle) {
+			return this.errorReply(`This command only works in battle rooms.`);
+		}
 		if (!battle.inputLog) {
 			this.errorReply(`This command only works when the battle has ended - if the battle has stalled, ask players to forfeit.`);
 			if (user.can('forcewin')) this.errorReply(`Alternatively, you can end the battle with /forcetie.`);
 			return;
 		}
 		if (!this.can('exportinputlog', null, room)) return;
-		if (!battle.allowExtraction) {
-			battle.allowExtraction = ['', ''];
-		}
-		if (user.userid === toID(battle.p1.name) || user.userid === toID(battle.p2.name)) {
-			this.parse(`/allowexportinputlog ${user.userid}`);
-		}
-		if (battle.allowExtraction[0] !== user.userid) {
-			const p1 = Users(battle.p1.name);
-			if (p1) p1.sendTo(room, Chat.html`|html|${user.name} wants to extract the battle input log. <button name="send" value="/allowexportinputlog ${user.userid}">Share your team and choices with "${user.name}"</button>`);
-		}
-		if (battle.allowExtraction[1] !== user.userid) {
-			const p2 = Users(battle.p2.name);
-			if (p2) p2.sendTo(room, Chat.html`|html|${user.name} wants to extract the battle input log. <button name="send" value="/allowexportinputlog ${user.userid}">Share your team and choices with "${user.name}"</button>`);
-		}
-
-		if (battle.allowExtraction.join(',') !== `${user.userid},${user.userid}`) {
-			this.addModAction(`${user.name} wants to extract the battle input log.`);
-			return;
+		if (!battle.allowExtraction[user.userid]) {
+			battle.allowExtraction[user.userid] = new Set();
+			for (const player of battle.players) {
+				const playerUser = Users(player.name);
```

Comment:
I would just do `player.getUser()` but it seems like `player.userid` is cleared when the battle ends, I'm not really sure why that's the case.

Reply:
Oh, it's the case because the user isn't considered a player when the battle ends (the user is allowed to change name without forfeiting, etc).

Reply:
It does make this code awkward, though. I wonder if there's a solution here.

Reply:
(This in general shouldn't be done because `player.name` isn't guaranteed to be a username; that's why `player.userid` exists in the first place.)

Reply:
Is this a better solution?
```javascript
const players = [battle.p1, battle.p2, battle.p3, battle.p4].filter(player => player);
for (const player of players) { ... }```

Reply:
> (This in general shouldn't be done because player.name isn't guaranteed to be a username; that's why player.userid exists in the first place.)

How so, is ``RoomBattlePlayer.name`` also not guaranteed to be a real name?

Reply:
No, none of those really deal with the core problem, which is relatively obscure.

One of them is `/importinputlog` in which case I think the player names can be whatever you want them to be?

This is probably fine as-is.

Reply:
I still don't understand why we clear `player.userid`.

Players can still change their name because the user rename methods check `this.games` which wouldn't contain `RoomBattle` because `unlinkUser` deletes it.

Reply:
Um, I want `player.userid` to stay in sync with `user.games`.

I guess that's not _strictly_ necessary though? I suppose I can accept a desync at the end of game, or, like, a flag or something.

Reply:
Anyway, yes, overall, I think that's a good idea. Let's just keep `player.userid` uncleared.

Reply:
Okay, I'll just make another PR for this, since it's a bit unrelated to this. I already have merge conflicts anyway..

Reply:
#5695.